### PR TITLE
[SPARK-20465][CORE] Throws a proper exception when any temp directory could not be got

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes to throw an exception with better message rather than `ArrayIndexOutOfBoundsException` when temp directories could not be created.

Running the commands below:

```bash
./bin/spark-shell --conf spark.local.dir=/NONEXISTENT_DIR_ONE,/NONEXISTENT_DIR_TWO
```

produces ...

**Before**

```
Exception in thread "main" java.lang.ExceptionInInitializerError
        ...
Caused by: java.lang.ArrayIndexOutOfBoundsException: 0
        ...
```

**After**

```
Exception in thread "main" java.lang.ExceptionInInitializerError
        ...
Caused by: java.io.IOException: Failed to get a temp directory under [/NONEXISTENT_DIR_ONE,/NONEXISTENT_DIR_TWO].
        ...
```

## How was this patch tested?

Unit tests in `LocalDirsSuite.scala`.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
This actually also happens when the directory exists but the user does not have the permission.

**Comment by HyukjinKwon** [comment]
@joshrosen, could you take a look and see if it makes sense?

File: core/src/test/scala/org/apache/spark/storage/LocalDirsSuite.scala [hunk 1]
```diff
@@ -35,20 +35,31 @@ class LocalDirsSuite extends SparkFunSuite with BeforeAndAfter {
...
 
+  test("Utils.getLocalDir() throws an exception if any temporary directory cannot be retrieved") {
+    assert(!new File("/NONEXISTENT_PATH_ONE").exists())
+    assert(!new File("/NONEXISTENT_PATH_TWO").exists())
+    val conf = new SparkConf(false)
+      .set("spark.local.dir", "/NONEXISTENT_PATH_ONE,/NONEXISTENT_PATH_TWO")
+    val message = intercept[IOException] {
+      Utils.getLocalDir(conf)
+    }.getMessage
+    assert(message ===
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/storage/LocalDirsSuite.scala]: I would only suggest not testing the exact message here, but it's not a big deal. Maybe testing for the existence of the paths.

**Review by srowen — APPROVED** [review]
Seems reasonable to me vs failing with a less clear exception

**Comment by HyukjinKwon** [comment]
Thank you @srowen.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Current status: I don't know how using

```scala
val dirs = Array.empty[Int]
dirs.headOption.getOrElse {
   throw new Exception("")
}
```

instead of 

```scala
val dirs = Array.empty[Int]
dirs(0)
```

could make the tests failed (given the observations above).
In the last commit, I gave a shot it with if-else as below:

```scala
val dirs = Array.empty[Int]
if (dirs.nonEmpty) {
  dirs(0)
} else {
  throw new Exception("")
}
```

To make sure logically this should not make other tests flaky.

**Comment by HyukjinKwon** [comment]
It looks we should clean up the cached local dirs after the test. I change it back to the original proposal with the cleanup.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by srowen** [comment]
Merged to master/2.2

**Pull request closed without merging**